strawberryfields/engine.py

```diff
@@ -609,15 +610,63 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         """
         # get the specific chip to submit the program to
         compile_options = compile_options or {}
+        recompile = compile_options.pop("recompile", False)
         kwargs.update(self._backend_options)
 
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        if program.compile_info is None:
+            # program is uncompiled
+            msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+            program = program.compile(device=device, **compile_options)
+
+        elif (
+            program.compile_info[0].target != device.target
+            or program.compile_info[0]._spec != device._spec
```

Comment:
What if a `DeviceSpec` is explicitly created so? The following snippet creates a `DeviceSpec` by specifying `X8` as target (instead of `X8_01`).

```python
import strawberryfields as sf
import strawberryfields.ops as ops
from strawberryfields.utils import random_interferometer
U = random_interferometer(4)

from strawberryfields.api import DeviceSpec

import textwrap

test_spec = {
    "layout": textwrap.dedent(
        """\
        name template_4x2_X8
        version 1.0
        target X8_01 (shots=1)

        # for n spatial degrees, first n signal modes, then n idler modes, all phases zero
        S2gate({squeezing_amplitude_0}, 0.0) | [0, 4]
        S2gate({squeezing_amplitude_1}, 0.0) | [1, 5]
        S2gate({squeezing_amplitude_2}, 0.0) | [2, 6]
        S2gate({squeezing_amplitude_3}, 0.0) | [3, 7]

        # standard 4x4 interferometer for the signal modes (the lower ones in frequency)
        # even phase indices correspond to internal Mach-Zehnder interferometer phases
        # odd phase indices correspond to external Mach-Zehnder interferometer phases
        MZgate({phase_0}, {phase_1}) | [0, 1]
        MZgate({phase_2}, {phase_3}) | [2, 3]
        MZgate({phase_4}, {phase_5}) | [1, 2]
        MZgate({phase_6}, {phase_7}) | [0, 1]
        MZgate({phase_8}, {phase_9}) | [2, 3]
        MZgate({phase_10}, {phase_11}) | [1, 2]

        # duplicate the interferometer for the idler modes (the higher ones in frequency)
        MZgate({phase_0}, {phase_1}) | [4, 5]
        MZgate({phase_2}, {phase_3}) | [6, 7]
        MZgate({phase_4}, {phase_5}) | [5, 6]
        MZgate({phase_6}, {phase_7}) | [4, 5]
        MZgate({phase_8}, {phase_9}) | [6, 7]
        MZgate({phase_10}, {phase_11}) | [5, 6]

        # add final dummy phases to allow mapping any unitary to this template (these do not
        # affect the photon number measurement)
        Rgate({final_phase_0}) | [0]
        Rgate({final_phase_1}) | [1]
        Rgate({final_phase_2}) | [2]
        Rgate({final_phase_3}) | [3]
        Rgate({final_phase_4}) | [4]
        Rgate({final_phase_5}) | [5]
        Rgate({final_phase_6}) | [6]
        Rgate({final_phase_7}) | [7]

        # measurement in Fock basis
        MeasureFock() | [0, 1, 2, 3, 4, 5, 6, 7]
    """
    ),
    "modes": 8,
    "compiler": [],
    "gate_parameters": {
        "squeezing_amplitude_0": [0, 1],
        "squeezing_amplitude_1": [0, 1],
        "squeezing_amplitude_2": [0, 1],
        "squeezing_amplitude_3": [0, 1],
        "phase_0": [0, [0, 6.283185307179586]],
        "phase_1": [0, [0, 6.283185307179586]],
        "phase_2": [0, [0, 6.283185307179586]],
        "phase_3": [0, [0, 6.283185307179586]],
        "phase_4": [0, [0, 6.283185307179586]],
        "phase_5": [0, [0, 6.283185307179586]],
        "phase_6": [0, [0, 6.283185307179586]],
        "phase_7": [0, [0, 6.283185307179586]],
        "phase_8": [0, [0, 6.283185307179586]],
        "phase_9": [0, [0, 6.283185307179586]],
        "phase_10": [0, [0, 6.283185307179586]],
        "phase_11": [0, [0, 6.283185307179586]],
        "final_phase_0": [0, [0, 6.283185307179586]],
        "final_phase_1": [0, [0, 6.283185307179586]],
        "final_phase_2": [0, [0, 6.283185307179586]],
        "final_phase_3": [0, [0, 6.283185307179586]],
        "final_phase_4": [0, [0, 6.283185307179586]],
        "final_phase_5": [0, [0, 6.283185307179586]],
        "final_phase_6": [0, [0, 6.283185307179586]],
        "final_phase_7": [0, [0, 6.283185307179586]],
    },
}

def generate_X8_params(r, p):
    return {
        "squeezing_amplitude_0": r,
        "squeezing_amplitude_1": r,
        "squeezing_amplitude_2": r,
        "squeezing_amplitude_3": r,
        "phase_0": p,
        "phase_1": p,
        "phase_2": p,
        "phase_3": p,
        "phase_4": p,
        "phase_5": p,
        "phase_6": p,
        "phase_7": p,
        "phase_8": p,
        "phase_9": p,
        "phase_10": p,
        "phase_11": p,
        "final_phase_0": 1.24,
        "final_phase_1": 0.54,
        "final_phase_2": 4.12,
        "final_phase_3": 0,
        "final_phase_4": 1.24,
        "final_phase_5": 0.54,
        "final_phase_6": 4.12,
        "final_phase_7": 0,
    }

X8_spec = DeviceSpec(target="X8", connection=None, spec=test_spec) # <--- specifying X8 as target instead of X8_01

params = generate_X8_params(1, 0.3)

prog = X8_spec.create_program(**params)

eng = sf.RemoteEngine("X8_01")
device = eng.device_spec

res = eng.run(prog, shots=10)
```

Reply:
:thinking: Good point @antalszava. That might actually be a case when these can differ, although I'm not sure if that's actually something that should be supported. Is there a reason _not_ to raise an error in the `DeviceSpec` class if someone attempts to create a device spec with a different target than the one defined in the layout?

I think that every time the engine is initiated, the `self.target` will always be a specific target (e.g. `X8_01`) rather than (e.g. `X8`). So, when calling the device spec API and creating the `DeviceSpec` object, there will never be a case where they differ.

Reply:
Thanks @antalszava. The only way this seems to be possible is if a user manually creates a device spec such that the target attribute does not match the created engine... This feels like something that should be validated (_if_ it is ever likely to happen) at a much lower level?

Reply:
> Is there a reason not to raise an error in the DeviceSpec class if someone attempts to create a device spec with a different target than the one defined in the layout?

That's a good question, not entirely sure myself either.

> This feels like something that should be validated (if it is ever likely to happen) at a much lower level?

What overall comes to mind is to remove the `target` argument from the `DeviceSpec` class and define `DeviceSpec._target` based on the `layout` that is being passed (by using the blackbird io function `loads` and extracting the target). This would ensure that `DeviceSpec._target` cannot differ from the target specified in `DeviceSpec._spec`. Could that potentially cause any major breaks? :thinking:

Any alternative validation methods would need to be included in `RemoteEngine.run`, as that is where `Program.target` and `RemoteEngine.target` can be compared.

Reply:
>  This would ensure that DeviceSpec._target cannot differ from the target specified in DeviceSpec._spec. Could that potentially cause any major breaks? 🤔

Hmm, it could potentially cause some breaks, so let's mark this to investigate in a separate PR